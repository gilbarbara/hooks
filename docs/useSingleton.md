# useSingleton

Run the code just once before the render.  
Similar to constructors in classes.

## Usage

```tsx
import { useRef } from 'react';
import ReactDOM from 'react-dom';
import { useSingleton } from '@gilbarbara/hooks';

function Component() {
  const node = useRef<HTMLElement | null>(null);

  useSingleton(() => {
    // this code will only be executed once.
    node.current = document.createElement('div');
    node.current.id = 'MyPortal';

    document.body.appendChild(node.current);
  });

  return node.current ? ReactDOM.createPortal(<div>Portal</div>, node.current) : null;
}
```

## Reference

```typescript
useSingleton(callback: () => void): void;
```
