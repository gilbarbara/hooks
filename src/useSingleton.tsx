import { useRef } from 'react';

export function useSingleton(callback: () => void): void {
  const hasBeenCalled = useRef(false);

  if (hasBeenCalled.current) {
    return;
  }

  callback();
  hasBeenCalled.current = true;
}
