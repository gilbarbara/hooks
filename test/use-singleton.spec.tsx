import * as React from 'react';
import { fireEvent, render } from '@testing-library/react';

import useSingleton from '../src/use-singleton';

const mockFn = jest.fn();

function Component() {
  const [count, setCount] = React.useState(0);

  useSingleton(() => {
    mockFn();
  });

  const handleClick = () => {
    setCount(s => s + 1);
  };

  return (
    <button onClick={handleClick} type="button">
      Count: {count}
    </button>
  );
}

describe('useSingleton', () => {
  afterEach(() => {
    mockFn.mockReset();
  });

  it('should call the singleton just once', () => {
    const { getByRole } = render(<Component />);

    expect(mockFn).toHaveBeenCalledTimes(1);

    fireEvent.click(getByRole('button'));
    fireEvent.click(getByRole('button'));
    fireEvent.click(getByRole('button'));
    fireEvent.click(getByRole('button'));

    expect(mockFn).toHaveBeenCalledTimes(1);
    expect(getByRole('button')).toMatchSnapshot();
  });
});
