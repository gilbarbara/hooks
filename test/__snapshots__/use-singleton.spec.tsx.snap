// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSingleton should call the singleton just once 1`] = `
<button
  type="button"
>
  Count: 
  4
</button>
`;
